pkg/internal/storage/repo/sync_state.go

```diff
@@ -0,0 +1,123 @@
+package repo
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/satoshi"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/database/models"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/database/scopes"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/internal/storage/entity"
+	"github.com/4chain-ag/go-wallet-toolbox/pkg/wdk"
+	"github.com/go-softwarelab/common/pkg/to"
+	"gorm.io/gorm"
+)
+
+type SyncState struct {
+	db *gorm.DB
+}
+
+func NewSyncState(db *gorm.DB) *SyncState {
+	return &SyncState{
+		db: db,
+	}
+}
+
+func (s *SyncState) FindSyncState(ctx context.Context, userID int, storageIdentityKey string) (*entity.SyncState, error) {
+	var model models.SyncState
+	err := s.db.WithContext(ctx).
+		Scopes(scopes.UserID(userID)).
+		Where("storage_identity_key = ?", storageIdentityKey).
+		First(&model).Error
+
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to find sync state: %w", err)
+	}
+
+	return mapModelToSyncStateEntity(model)
+}
+
+func (s *SyncState) CreateSyncState(ctx context.Context, syncState *entity.SyncState) (*entity.SyncState, error) {
+	syncMapJSON, err := syncState.SyncMap.JSON()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal sync map: %w", err)
+	}
+
+	model := models.SyncState{
+		UserID:             syncState.UserID,
+		StorageIdentityKey: syncState.StorageIdentityKey,
+		StorageName:        syncState.StorageName,
+		Status:             syncState.Status,
+		RefNum:             syncState.Reference,
+		SyncMap:            syncMapJSON,
+		When:               syncState.When,
+	}
+
+	if syncState.Satoshis != nil {
+		model.Satoshis = to.Ptr(syncState.Satoshis.Int64())
+	}
+
+	err = s.db.WithContext(ctx).Create(&model).Error
+	if err != nil {
+		return nil, fmt.Errorf("failed to create sync state: %w", err)
+	}
+
+	return mapModelToSyncStateEntity(model)
+}
+
+func (s *SyncState) UpdateSyncState(ctx context.Context, syncState *entity.SyncState) error {
+	syncMapJSON, err := syncState.SyncMap.JSON()
+	if err != nil {
+		return fmt.Errorf("failed to marshal sync map: %w", err)
+	}
+
+	toUpdate := models.SyncState{
+		Status:  syncState.Status,
+		SyncMap: syncMapJSON,
+		When:    syncState.When,
+	}
+
+	if syncState.Satoshis != nil {
+		toUpdate.Satoshis = to.Ptr(syncState.Satoshis.Int64())
+	}
```

Comment:
```suggestion
	toUpdate := map[string]interface{}{
    "status":  syncState.Status,
    "sync_map": syncMapJSON,
    "when":    syncState.When,
}

if syncState.Satoshis != nil {
    toUpdate["satoshis"] = syncState.Satoshis.Int64()
}
```

would not be better to use map here instead of struct ? since i think GORM works better with maps as only updates what u provided ... no risk of zero overwrite ? 
https://gorm.io/docs/update.html#Update-Changed-Fields

Reply:
Sure, why not. 
But there is one drawback of this method, (just to note) - that we need to use "db column names" - so the golang compiler can't check if we use proper column names 🤷